### PR TITLE
Proximity forest

### PR DESCRIPTION
**Pull request by goastler**
Implementation of Proximity Forest. I have verified it is fully working on two datasets, however we may want to wait for others to complete before calling this done / merging this pull request.

Proximity Forest has 3 classes:

- Proximity stump: essentially a 'split' as referred to in the proximity forest paper. Simply a 1 nearest neighbour classifier with n exemplar train instances (where n is the number of classes in the problem).
- Proximity tree: a classic decision tree. Each tree contains a proximity stump which decides the proximity of instances to a set of exemplars. A new branch is created for each exemplar. Each branch has a sub-tree which is fed the instances closest to the exemplar, as determined in the proximity stump. This continues until nodes are pure, as per the paper.
- Proximity forest: a forest of proximity trees. Simple to understand, it's an ensemble of trees with majority vote.

I have designed it to have callable methods as parameters where possible, allowing users to define how to obtain exemplars, how to calculate gain, etc.

Differences to the published work:
- Does not include TW

I've made a series of generic methods in the utilities class. These find mins / maxs / best values in arrays and differ to numpy in using random tie breaks. There's also a few helper classes in there (e.g. population std dev). There's a boilerplate classifier class in classifier.py which implemented majority prediction from predict_proba.

### CODE REVIEW

**Comment by goastler** [comment]
need to make sure it works with orchestration class

**Comment by goastler** [comment]
how should dimensionality be handled?

File: sktime/classifiers/proximity.py [hunk 1]
```diff
@@ -0,0 +1,1084 @@
...
+            sq_proportion = np.math.pow(proportion, 2)
+            score -= sq_proportion
+    # double score as gini is between 0 and 0.5, we need 0 and 1
+    score *= 2
+    return score
+
+
+# todo info gain
+def information_gain(parent_class_labels, children_class_labels):
+    raise Exception('not implemented yet')
```
Comment (mloning) [inline sktime/classifiers/proximity.py]:
```suggestion
raise NotImplementedError()
```

File: sktime/classifiers/proximity.py [hunk 2]
```diff
@@ -0,0 +1,1084 @@
...
+
+
+# todo chi sq
+def chi_squared(parent_class_labels, children_class_labels):
+    raise Exception('not implemented yet')
```
Comment (mloning) [inline sktime/classifiers/proximity.py]:
```suggestion
raise NotImplementedError()
```

File: sktime/classifiers/proximity.py [hunk 3]
```diff
@@ -0,0 +1,1084 @@
...
+
+
+def pick_one_exemplar_per_class(instances, class_labels, rand):
```
Comment (mloning) [inline sktime/classifiers/proximity.py]: For handling the random number generation here and throughout, I would rely on sklearn's [`check_random_state()`](http://scikit-learn.org/stable/modules/generated/sklearn.utils.check_random_state.html)

File: sktime/classifiers/proximity.py [hunk 4]
```diff
@@ -0,0 +1,1084 @@
...
+        children_score_sum += child_score
+    # gini outputs relative improvement
+    score = parent_score - children_score_sum
+    return score
+
+
+def gini_node(class_labels):
+    '''
+    get gini score at a specific node
+    ----
```
Comment (mloning) [inline sktime/classifiers/proximity.py]:
```suggestion
```
That line is not necessary in the docstring I think

File: sktime/classifiers/proximity.py [hunk 5]
```diff
@@ -0,0 +1,1084 @@
...
+    '''
+
+    __author__ = 'George Oastler (linkedin.com/goastler)'
+
+    def __init__(self,
+                 pick_exemplars_method = None,
+                 param_perm = None,
+                 gain_method = None,
+                 label_encoder = None,
+                 rand = np.random.RandomState):
```
Comment (mloning) [inline sktime/classifiers/proximity.py]: Yes, so here you could just have `None` as the default and then pass that to `check_random_state()`

File: sktime/classifiers/proximity.py [hunk 6]
```diff
@@ -0,0 +1,1084 @@
...
+        param_perm constructor parameter.
+        ----
+        Returns
+        ----
+        key : string
+            key for the distance measure for the param_perm dict
+        '''
+        return 'dm'
+
+    def fit(self, instances, class_labels, should_check_data = True):
```
Comment (mloning) [inline sktime/classifiers/proximity.py]: To make it consistent, rename: 
* instances to X
* class_labels to y
* should_check_data to input_checks

File: sktime/classifiers/proximity.py [hunk 7]
```diff
@@ -0,0 +1,1084 @@
...
+    param_pool : list of dicts
+        a list of dictionaries containing a distance measure and corresponding parameter sources (distribution or
+        predefined value)
+    gain_method : callable
+        a method to calculate the gain of this split / stump
+    label_encoder : LabelEncoder
+        a label encoder, can be pre-populated
+    rand : numpy RandomState
+        a random state for sampling random numbers
+    r : int
```
Comment (mloning) [inline sktime/classifiers/proximity.py]: Maybe rename r to n_stumps?

File: sktime/utils/classifier.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
+import numpy as np
+from sklearn.base import BaseEstimator, ClassifierMixin
+
+from sktime.utils import utilities
+from sktime.utils.utilities import check_data
+
+
+class Classifier(BaseEstimator, ClassifierMixin):
```
Comment (mloning) [inline sktime/utils/classifier.py]: Please inherit from the already existing [`BaseClassifier`](https://github.com/alan-turing-institute/sktime/blob/master/sktime/classifiers/base.py), but you can of course refactor your class into a `BaseProximityClassifier` or something which inherits from `BaseClassifier`

File: sktime/utils/utilities.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
+import numpy as np
+from pandas import DataFrame
```
Comment (mloning) [inline sktime/utils/utilities.py]: With `pandas`, I would always use `import pandas as pd` and then `pd.DataFrame`, that seems to be the convention

File: sktime/utils/utilities.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+# little accuracy helper function
+def accuracy(actual_class_values, predicted_class_values):
```
Comment (mloning) [inline sktime/utils/utilities.py]: There is an `accuracy_score` function in sklearn which you can use

**Comment by goastler** [comment]
Cheers @mloning , I'll make those changes now. 

My only reservation is with comment 1 about moving the utility methods. I agree for the specific ones (e.g. finding the pop. std dev of a dataset) but disagree with the more generic ones (e.g. arg_best, min, max, best, etc) as these can easily be used in other stuff, therefore I don't think they should go in the proximity forest package as they're unspecific to proximity forest. Having said that, I have no idea where to put them otherwise. A good example is that I'll probably use them in the kernel work, so need to be in an appropriate place

Perhaps leaving them in the utils package would be best (so they're shared amongst everything) but rename to file to comparison.py (working title) as most of the generic stuff is to compare array elements in some way. Would that suffice?

File: sktime/utils/utilities.py (hunk 10 above)
Comment (goastler) [reply sktime/utils/utilities.py]: ah this was before I discovered that, I'll get rid of it as it's redundant now

File: sktime/utils/classifier.py (hunk 8 above)
Comment (goastler) [reply sktime/utils/classifier.py]: Can I add the predict method to `BaseClassifier`? The predict method is usually the same for every classifier

File: sktime/classifiers/proximity.py (hunk 7 above)
Comment (goastler) [reply sktime/classifiers/proximity.py]: It's called `r` in the paper and Java code so I simply mimicked them. I think `n_stumps` is better, what do you think @jasonlines @TonyBagnall ?

File: sktime/classifiers/proximity.py (hunk 3 above)
Comment (goastler) [reply sktime/classifiers/proximity.py]: Ah great method, thanks for the pointer! :)

**Comment by goastler** [comment]
Re callables as parameters, sklearn seems to do it all over the place so I'll leave PF params as is

File: sktime/utils/classifier.py (hunk 8 above)
Comment (mloning) [reply sktime/utils/classifier.py]: Currently, all of the classifiers have their own predict methods as far as I know, so adding this without deleting the class specific ones won't do anything. Refactoring out the `predict` method seems sensible though, but not sure if we want to do this now, I guess we would have to do some testing again or write  more extensive unit tests first.

Comment (goastler) [reply sktime/utils/classifier.py]: Seems more hassle than it's worth, I've just moved the predict method to a static method and made classifiers call that, which still reduces ~15 lines to 1

**Comment by goastler** [comment]
That should be all comments / issues address!

**Pull request merged**